From 738df431307b7f28a9573cf63a837569a39c7a05 Mon Sep 17 00:00:00 2001
From: Longfang Liu <liulongfang@huawei.com>
Date: Mon, 11 Sep 2023 18:22:16 +0800
Subject: [PATCH 571/960] downstream: xhci: fix USB xhci controller issue

When the current HiSilicon USB xhci controller formats the faulty
U disk, it will trigger a controller exception error. This will
cause errors in the control logic of the xhci controller and
driver software. In the end, all USB devices on the xhci controller
cannot be used.

By introducing a noop command operation, restore the logic of the
xhci controller and driver software, and restore all USB devices
on the xhci controller to normal.

Signed-off-by: Longfang Liu <liulongfang@huawei.com>
Signed-off-by: JiangShui Yang <yangjiangshui@h-partners.com>
Signed-off-by: Slim6882 <15605196882@163.com>
Signed-off-by: chenyi <chenyi211@h-partners.com>
Signed-off-by: Liu Shixin <liushixin2@huawei.com>
---
 drivers/usb/host/xhci-pci.c  | 4 ++++
 drivers/usb/host/xhci-ring.c | 6 +++++-
 drivers/usb/host/xhci.h      | 1 +
 3 files changed, 10 insertions(+), 1 deletion(-)

diff --git a/drivers/usb/host/xhci-pci.c b/drivers/usb/host/xhci-pci.c
index 5abc48f148dc..5f4e3b9075d8 100644
--- a/drivers/usb/host/xhci-pci.c
+++ b/drivers/usb/host/xhci-pci.c
@@ -576,6 +576,10 @@ static void xhci_pci_quirks(struct device *dev, struct xhci_hcd *xhci)
 	if (xhci->quirks & XHCI_RESET_ON_RESUME)
 		xhci_dbg_trace(xhci, trace_xhci_dbg_quirks,
 				"QUIRK: Resetting on resume");
+
+	if (pdev->vendor == PCI_VENDOR_ID_HUAWEI &&
+	   (pdev->device == 0xa23c || pdev->device == 0xa23d))
+		xhci->quirks |= XHCI_USB3_NOOP;
 }
 
 #ifdef CONFIG_ACPI
diff --git a/drivers/usb/host/xhci-ring.c b/drivers/usb/host/xhci-ring.c
index cdb819e323b3..1466aaf1f6a5 100644
--- a/drivers/usb/host/xhci-ring.c
+++ b/drivers/usb/host/xhci-ring.c
@@ -1813,7 +1813,9 @@ static void handle_cmd_completion(struct xhci_hcd *xhci,
 	 * Check whether the completion event is for our internal kept
 	 * command.
 	 */
-	if (!cmd_dequeue_dma || cmd_dma != (u64)cmd_dequeue_dma) {
+	if (!cmd_dequeue_dma || ((cmd_dma != (u64)cmd_dequeue_dma) &&
+	    !((xhci->quirks & XHCI_USB3_NOOP) && (cmd_comp_code ==
+	       COMP_COMMAND_RING_STOPPED)))) {
 		xhci_warn(xhci,
 			  "ERROR mismatched command completion event\n");
 		return;
@@ -1838,6 +1840,8 @@ static void handle_cmd_completion(struct xhci_hcd *xhci,
 	if (cmd_comp_code == COMP_COMMAND_ABORTED) {
 		xhci->cmd_ring_state = CMD_RING_STATE_STOPPED;
 		if (cmd->status == COMP_COMMAND_ABORTED) {
+			if (xhci->quirks & XHCI_USB3_NOOP)
+				trb_to_noop(cmd->command_trb, TRB_CMD_NOOP);
 			if (xhci->current_cmd == cmd)
 				xhci->current_cmd = NULL;
 			goto event_handled;
diff --git a/drivers/usb/host/xhci.h b/drivers/usb/host/xhci.h
index 849a83e4013c..590728f810e0 100644
--- a/drivers/usb/host/xhci.h
+++ b/drivers/usb/host/xhci.h
@@ -1598,6 +1598,7 @@ struct xhci_hcd {
 #define XHCI_STATE_HALTED	(1 << 1)
 #define XHCI_STATE_REMOVING	(1 << 2)
 	unsigned long long	quirks;
+#define XHCI_USB3_NOOP BIT_ULL(63)
 #define	XHCI_LINK_TRB_QUIRK	BIT_ULL(0)
 #define XHCI_RESET_EP_QUIRK	BIT_ULL(1) /* Deprecated */
 #define XHCI_NEC_HOST		BIT_ULL(2)
-- 
2.43.0

